pkg/controller/shared_state.go

```diff
@@ -1345,13 +1371,160 @@ func (c *SharedState) createReplicas(ctx context.Context, remainingReplicas int,
 		//no failed replica attachments, but there are still more replicas to reach MaxShares
 		request := ReplicaRequest{VolumeName: volumeName, Priority: remainingReplicas}
 		c.priorityReplicaRequestsQueue.Push(ctx, &request)
-		for _, pod := range pods {
-			c.eventRecorder.Eventf(pod.DeepCopyObject(), v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, "Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		message := fmt.Sprintf("Not enough suitable nodes to attach %d of %d replica mount(s) for volume %s", remainingReplicas, requiredReplicas, volumeName)
+		podCopies := make([]runtime.Object, len(pods))
+		for i, pod := range pods {
+			podCopies[i] = pod.DeepCopyObject()
 		}
+		timestamp := time.Now()
+		for _, podCopy := range podCopies {
+			c.eventRecorder.Eventf(podCopy, v1.EventTypeWarning, consts.ReplicaAttachmentFailedEvent, message)
+		}
+		message += timestamp.UTC().Format(" [0102 15:04:05") // append original timestamp
+		c.persistAttachmentFailure(volumeName, message, podCopies, timestamp)
 	}
 	return nil
 }
 
+func (c *SharedState) persistAttachmentFailure(volumeName string, message string, podCopies []runtime.Object, timestamp time.Time) {
+	c._eventRefresherInitOnce.Do(c._eventRefresherInitFunc)
+	c.eventsToPersistQueue <- ReplicaAttachmentFailureInfo{volumeName, message, podCopies, timestamp}
+}
+
+// Only requests an event un-persist if the event refresher is active
+func (c *SharedState) unpersistAttachmentFailure(volumeName string) {
+	trySend(c.eventsToUnpersistQueue, volumeName) // also mitigates unlikely race condition of sending to closed channel
+}
+
+func (c *SharedState) _eventRefresherInitFunc() {
+	c.eventsToPersistQueue = make(chan ReplicaAttachmentFailureInfo, maxPersistEventConcurrency)
```

Comment:
Since there is only a single goroutine processing event persist/unpersist notifications, the concurrency constants control concurrency of production rather than consumption. This means that if persist/unpersist events are produced at a faster pace than the goroutine can process them the producers will block. This could provide to be an unnecessary serialization point on the producer side. I believe we can choose an arbitrarily high buffer size. IIRC, this is just a max count of queued items before blocking the producer and doesn't result in any real allocation.

Reply:
You could probably make the queue length equal to the number of controller [`WorkerThreads`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/ce8a4cf7054558f2a1ac68a9f47a6ab0f1432a24/pkg/apis/azuredisk/v1beta2/types.go#L575C6-L575C6).